File: base/src/test/java/com/mindera/skeletoid/logs/appenders/LogFileAppenderUnitTest.java
package com.mindera.skeletoid.logs.appenders;

import android.content.Context;
import android.content.Intent;
import android.support.v4.content.FileProvider;

import com.mindera.skeletoid.generic.AndroidUtils;
import com.mindera.skeletoid.logs.LOG;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import java.io.File;
import java.util.logging.FileHandler;
import java.util.logging.Level;
import java.util.logging.LogRecord;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(PowerMockRunner.class)
@PrepareForTest({FileProvider.class, Intent.class, AndroidUtils.class})
public class LogFileAppenderUnitTest {

    private String PACKAGE_NAME = "my.package.name";
    private final String FILE_NAME = "FILENAME";


    @Test(expected = IllegalArgumentException.class)
    public void testConstructorAllNull() {
        new LogFileAppender(null, null);

    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructorPackageNameNull() {
        new LogFileAppender(null, FILE_NAME);

    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructorFileNameInvalid() {
        new LogFileAppender(PACKAGE_NAME, "*");

    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructorFileNameNull() {
        new LogFileAppender(PACKAGE_NAME, null);

    }

    @Test
    public void testFileNameIsValidEmpty() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertFalse(appender.isFilenameValid(""));
    }

    @Test
    public void testFileNameIsValidInvalid() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertFalse(appender.isFilenameValid("//"));
    }

    @Test
    public void testFileNameIsValid() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertTrue(appender.isFilenameValid(FILE_NAME));
    }


    @Test
    public void testConstructor() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertEquals("LogFileAppender", appender.getLoggerId());
    }

    @Test
    public void testEnableAppender() {
        Context context = mock(Context.class);
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        appender.enableAppender(context);

        assertFalse(appender.canWriteToFile());
    }

    @Test
    public void testDisableAppender() throws InterruptedException {
        Context context = mock(Context.class);
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);
        appender.enableAppender(context);
        appender.disableAppender();
        Thread.sleep(600);

        assertFalse(appender.canWriteToFile());
    }

    @Test
    public void testCanWriteToFile() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertFalse(appender.canWriteToFile());
    }

    @Test
    public void testFormatLog() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertTrue(appender.formatLog(LOG.PRIORITY.DEBUG, null, "Hello", "My friend").matches("\\d\\d\\d\\d-\\d\\d-\\d\\d \\d\\d:\\d\\d:\\d\\d: D/" + PACKAGE_NAME + "\\(" + Thread.currentThread().getId() + "\\): Hello My friend "));
    }

    @Test
    public void testSetMaxLineLength() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        appender.setLogFileSize(1000);

        assertEquals(1000, appender.getLogFileSize());
    }

    @Test
    public void testSetNumberOfLogFiles() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        appender.setNumberOfLogFiles(5);

        assertEquals(5, appender.getNumberOfLogFiles());
    }

    @Test
    public void testSetMinLogLevel() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        appender.setMinLogLevel(LOG.PRIORITY.DEBUG);

        assertEquals(LOG.PRIORITY.DEBUG, appender.getMinLogLevel());
    }

    @Test
    public void testGetFileLogPath() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);
        Context context = mock(Context.class);

        File file = mock(File.class);
        when(context.getFilesDir()).thenReturn(file);
        when(file.getPath()).thenReturn("/com/mindera/skeletoid");

        assertEquals("/com/mindera/skeletoid/FILENAME.log", appender.getFileLogPath(context));
    }

    @Test
    public void testGetExternalFileLogPath() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME, true);
        Context context = mock(Context.class);
        PowerMockito.mockStatic(AndroidUtils.class);
        when(AndroidUtils.getExternalPublicDirectory(String.format("%s%s.log", File.separator, FILE_NAME))).thenReturn("external/com/mindera/skeletoid/FILENAME.log");

        assertEquals("external/com/mindera/skeletoid/FILENAME.log", appender.getFileLogPath(context));
    }

    @Test
    public void testFileHandlerLevelDebug() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertEquals(Level.SEVERE, appender.getFileHandlerLevel(LOG.PRIORITY.ERROR));
    }

    @Test
    public void testFileHandlerLevelFatal() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertEquals(Level.SEVERE, appender.getFileHandlerLevel(LOG.PRIORITY.FATAL));
    }

    @Test
    public void testFileHandlerLevelInfo() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertEquals(Level.INFO, appender.getFileHandlerLevel(LOG.PRIORITY.INFO));
    }

    @Test
    public void testFileHandlerLevelVerbose() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertEquals(Level.ALL, appender.getFileHandlerLevel(LOG.PRIORITY.VERBOSE));
    }

    @Test
    public void testFileHandlerLevelWarn() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertEquals(Level.WARNING, appender.getFileHandlerLevel(LOG.PRIORITY.WARN));
    }

    @Test
    public void testIsThreadPoolRunningWhenAppenderEnabled() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);
        Context context = mock(Context.class);
        appender.enableAppender(context);

        assertTrue(appender.isThreadPoolRunning());
    }

    @Test
    public void testIsThreadPoolNotRunningWhenAppenderInitialised() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);

        assertFalse(appender.isThreadPoolRunning());
    }

    @Test
    public void testIsThreadPoolNotRunningWhenAppenderDisabled() {
        Context context = mock(Context.class);
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);
        appender.enableAppender(context);

        appender.disableAppender();

        assertFalse(appender.isThreadPoolRunning());
    }

    @Test
    public void testLog() throws InterruptedException {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);
        FileHandler fileHandler = mock(FileHandler.class);
        Context context = mock(Context.class);
        File file = mock(File.class);
        when(context.getFilesDir()).thenReturn(file);
        when(file.getPath()).thenReturn("/com/mindera/skeletoid");
        appender.enableAppender(context);
        Thread.sleep(500);
        appender.mCanWriteToFile = true;
        appender.mFileHandler = fileHandler;

        appender.log(LOG.PRIORITY.DEBUG, new Throwable("oops"), "hello");

        verify(fileHandler).publish(any(LogRecord.class));
    }

    @Test
    public void testNoLogIfWritingNotEnabled() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);
        FileHandler fileHandler = mock(FileHandler.class);
        Context context = mock(Context.class);
        File file = mock(File.class);
        when(context.getFilesDir()).thenReturn(file);
        when(file.getPath()).thenReturn("/com/mindera/skeletoid");

        appender.log(LOG.PRIORITY.DEBUG, null, "hello");

        verify(fileHandler, times(0)).publish(any(LogRecord.class));
    }

    @Test
    public void testNoLogIfAppenderNotEnabled() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);
        FileHandler fileHandler = mock(FileHandler.class);

        appender.log(LOG.PRIORITY.DEBUG, null, "hello");

        verify(fileHandler, times(0)).publish(any(LogRecord.class));
    }

    @Test
    public void testNoLogIfAppenderDisabled() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);
        FileHandler fileHandler = mock(FileHandler.class);
        Context context = mock(Context.class);
        File file = mock(File.class);
        when(context.getFilesDir()).thenReturn(file);
        when(file.getPath()).thenReturn("/com/mindera/skeletoid");
        appender.enableAppender(context);
        appender.mCanWriteToFile = true;
        appender.mFileHandler = fileHandler;
        appender.disableAppender();

        appender.log(LOG.PRIORITY.DEBUG, null, "hello");

        verify(fileHandler, times(0)).publish(any(LogRecord.class));
    }

    @Test
    public void testNoLogIfLogLevelBelowMinLogLevel() {
        LogFileAppender appender = new LogFileAppender(PACKAGE_NAME, FILE_NAME);
        appender.setMinLogLevel(LOG.PRIORITY.ERROR);
        FileHandler fileHandler = mock(FileHandler.class);
        Context context = mock(Context.class);
        File file = mock(File.class);
        when(context.getFilesDir()).thenReturn(file);
        when(file.getPath()).thenReturn("/com/mindera/skeletoid");
        appender.enableAppender(context);
        appender.mFileHandler = fileHandler;

        appender.log(LOG.PRIORITY.DEBUG, null, "hello");

        verify(fileHandler, times(0)).publish(any(LogRecord.class));
    }
}

File: base/src/test/java/com/mindera/skeletoid/logs/appenders/LogFileAppenderUnitTest.kt
package com.mindera.skeletoid.logs.appenders

import android.content.Intent
import androidx.core.content.FileProvider
import com.mindera.skeletoid.generic.AndroidUtils
import com.mindera.skeletoid.logs.LOG
import org.junit.Ignore
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyNoMoreInteractions
import org.powermock.api.mockito.PowerMockito
import org.powermock.core.classloader.annotations.PowerMockIgnore
import org.powermock.core.classloader.annotations.PrepareForTest
import org.powermock.modules.junit4.rule.PowerMockRule
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import org.robolectric.annotation.Config
import java.io.File
import java.util.logging.FileHandler
import java.util.logging.Level
import java.util.logging.LogRecord
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE)
@PowerMockIgnore("org.mockito.*", "org.robolectric.*", "android.*")
@PrepareForTest(FileProvider::class, Intent::class, AndroidUtils::class)
public class LogFileAppenderUnitTest {

    companion object {
        private const val PACKAGE_NAME = "my.package.name"
        private const val FILE_NAME = "FILENAME"
    }

    @Rule
    @JvmField
    public var rule = PowerMockRule()

    @Test(expected = IllegalArgumentException::class)
    fun testConstructorFileNameInvalid() {
        LogFileAppender(PACKAGE_NAME, "*")
    }

    @Test
    fun testFileNameIsValidEmpty() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertFalse(appender.isFilenameValid(""))
    }

    @Test
    fun testFileNameIsValidInvalid() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertFalse(appender.isFilenameValid("//"))
    }

    @Test
    fun testFileNameIsValid() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertTrue(appender.isFilenameValid(FILE_NAME))
    }

    @Test
    fun testConstructor() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertEquals("LogFileAppender", appender.loggerId)
    }

    @Test
    fun testEnableAppender() {
        val context = RuntimeEnvironment.application
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        appender.enableAppender(context)
        Thread.sleep(2000)

        assertTrue(appender.canWriteToFile())
    }

    @Test
    fun testDisableAppender() {
        val context = RuntimeEnvironment.application
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)
        appender.enableAppender(context)
        Thread.sleep(2000)
        appender.disableAppender()

        assertFalse(appender.canWriteToFile())
    }

    @Test
    fun testCanWriteToFile() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertFalse(appender.canWriteToFile())
    }

    @Test
    fun testFormatLog() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertTrue(
            appender.formatLog(
                LOG.PRIORITY.DEBUG,
                "",
                "Hello",
                "My friend"
            ).matches(("\\d\\d\\d\\d-\\d\\d-\\d\\d \\d\\d:\\d\\d:\\d\\d: D/" + PACKAGE_NAME + "\\(" + Thread.currentThread().id + "\\): Hello My friend ").toRegex())
        )
    }

    @Test
    fun testSetMaxLineLength() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        appender.logFileSize = 1000

        assertEquals(1000, appender.logFileSize)
    }

    @Test
    fun testSetNumberOfLogFiles() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        appender.numberOfLogFiles = 5

        assertEquals(5, appender.numberOfLogFiles)
    }

    @Test
    fun testSetMinLogLevel() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        appender.minLogLevel = LOG.PRIORITY.DEBUG

        assertEquals(LOG.PRIORITY.DEBUG, appender.minLogLevel)
    }

    @Test
    fun testGetFileLogPath() {
        val context = RuntimeEnvironment.application
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)
        PowerMockito.mockStatic(AndroidUtils::class.java)
        `when`(
            AndroidUtils.getFileDirPath(
                context,
                String.format(
                    "%s%s.log",
                    File.separator,
                    FILE_NAME
                )
            )
        ).thenReturn("internal/com/mindera/skeletoid/FILENAME.log")
        assertEquals(
            "internal/com/mindera/skeletoid/FILENAME.log",
            appender.getFileLogPath(context)
        )
    }

    @Test
    fun testGetExternalFileLogPath() {
        val context = RuntimeEnvironment.application
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME, true)
        PowerMockito.mockStatic(AndroidUtils::class.java)
        `when`(
            AndroidUtils.getExternalPublicDirectory(
                String.format(
                    "%s%s.log",
                    File.separator,
                    FILE_NAME
                )
            )
        ).thenReturn("external/com/mindera/skeletoid/FILENAME.log")

        assertEquals(
            "external/com/mindera/skeletoid/FILENAME.log",
            appender.getFileLogPath(context)
        )
    }

    @Test
    fun testFileHandlerLevelDebug() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertEquals(Level.SEVERE, appender.getFileHandlerLevel(LOG.PRIORITY.ERROR))
    }

    @Test
    fun testFileHandlerLevelFatal() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertEquals(Level.SEVERE, appender.getFileHandlerLevel(LOG.PRIORITY.FATAL))
    }

    @Test
    fun testFileHandlerLevelInfo() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertEquals(Level.INFO, appender.getFileHandlerLevel(LOG.PRIORITY.INFO))
    }

    @Test
    fun testFileHandlerLevelVerbose() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertEquals(Level.ALL, appender.getFileHandlerLevel(LOG.PRIORITY.VERBOSE))
    }

    @Test
    fun testFileHandlerLevelWarn() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertEquals(Level.WARNING, appender.getFileHandlerLevel(LOG.PRIORITY.WARN))
    }

    @Test
    fun testIsThreadPoolRunningWhenAppenderEnabled() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)
        val context = RuntimeEnvironment.application

        appender.enableAppender(context)

        assertTrue(appender.isThreadPoolRunning)
    }

    @Test
    fun testIsThreadPoolNotRunningWhenAppenderInitialised() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)

        assertFalse(appender.isThreadPoolRunning)
    }

    @Test
    fun testIsThreadPoolNotRunningWhenAppenderDisabled() {
        val context = RuntimeEnvironment.application
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)
        appender.enableAppender(context)

        appender.disableAppender()

        assertFalse(appender.isThreadPoolRunning)
    }

    @Test
    @Ignore
    fun testLog() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)
        val context = RuntimeEnvironment.application
        val fileHandler = mock(FileHandler::class.java)

        appender.enableAppender(context)
        Thread.sleep(5000)
        appender.fileHandler = fileHandler

        appender.log(LOG.PRIORITY.DEBUG, Throwable("oops"), "hello")

        verify(fileHandler).publish(any(LogRecord::class.java))
    }

    @Test
    fun testNoLogIfWritingNotEnabled() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)
        val fileHandler = mock(FileHandler::class.java)
        appender.fileHandler = fileHandler

        appender.log(LOG.PRIORITY.DEBUG, null, "hello")

        verifyNoMoreInteractions(fileHandler)
    }

    @Test
    fun testNoLogIfAppenderNotEnabled() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)
        val fileHandler = mock(FileHandler::class.java)
        appender.fileHandler = fileHandler

        appender.log(LOG.PRIORITY.DEBUG, null, "hello")

        verifyNoMoreInteractions(fileHandler)
    }

    @Test
    fun testNoLogIfAppenderDisabled() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)
        val fileHandler = mock(FileHandler::class.java)
        val context = RuntimeEnvironment.application
        appender.enableAppender(context)
        Thread.sleep(2000)
        appender.fileHandler = fileHandler
        appender.disableAppender()

        appender.log(LOG.PRIORITY.DEBUG, null, "hello")

        verify(fileHandler, times(0)).publish(any(LogRecord::class.java))
    }

    @Test
    fun testNoLogIfLogLevelBelowMinLogLevel() {
        val appender = LogFileAppender(PACKAGE_NAME, FILE_NAME)
        appender.minLogLevel = LOG.PRIORITY.ERROR
        val fileHandler = mock(FileHandler::class.java)
        val context = RuntimeEnvironment.application
        appender.enableAppender(context)
        Thread.sleep(2000)
        appender.fileHandler = fileHandler

        appender.log(LOG.PRIORITY.DEBUG, null, "hello")

        verifyNoMoreInteractions(fileHandler)
    }
}
